components/kyma-adapter/internal/gqlclient/gqlclient.go

```diff
@@ -0,0 +1,219 @@
+package gqlclient
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/log"
+
+	"github.com/avast/retry-go/v4"
+	"github.com/kyma-incubator/compass/components/director/pkg/graphql"
+	gcli "github.com/machinebox/graphql"
+	"github.com/pkg/errors"
+)
+
+var tenantHeader = "tenant"
+
+// NewClient constructs a client
+func NewClient(gqlClient *gcli.Client) *client {
+	return &client{Client: gqlClient}
+}
+
+// Client is extended graphql.Client with custom requests
+type client struct {
+	*gcli.Client
+}
+
+type gqlResult struct {
+	Result interface{} `json:"result"`
+}
+
+// ApplicationBundles contains Bundle page for an application
+type ApplicationBundles struct {
+	Bundles graphql.BundlePageExt `json:"bundles"`
+}
+
+// runWithTenant executes gql request with tenant header and with retry on connectivity problems
+func (c *client) runWithTenant(ctx context.Context, gqlReq *gcli.Request, tenant string, resp interface{}) error {
+	gqlReq.Header.Set(tenantHeader, tenant)
+
+	return withRetryOnTemporaryConnectionProblems(ctx, func() error {
+		return c.Client.Run(ctx, gqlReq, resp)
+	})
+}
+
+// GetApplicationBundles gets all bundles for an application with appID and tenant using the internal gql client
+func (c *client) GetApplicationBundles(ctx context.Context, appID, tenant string) ([]*graphql.BundleExt, error) {
+	strReq := `query {
+  			result: application(id: "%s") {
+  			  bundles(first:%d, after:"%s") {
+  			    data {
+  			      id
+				  instanceAuths {
+				    id
+				    runtimeID
+				  }
+  			    }
+  			    pageInfo {
+  			      hasNextPage
+  			      endCursor
+  			    }
+  			    totalCount
+  			  }
+  			}
+		}`
+	pageSize := 200
+	initialGqlReq := gcli.NewRequest(fmt.Sprintf(strReq, appID, pageSize, ""))
+
+	appBundles := ApplicationBundles{}
+	gqlRes := gqlResult{Result: &appBundles}
+
+	if err := c.runWithTenant(ctx, initialGqlReq, tenant, &gqlRes); err != nil {
+		return nil, errors.Wrapf(err, "Error while getting bundles for application with id %q", appID)
+	}
+
+	result := make([]*graphql.BundleExt, 0, appBundles.Bundles.TotalCount)
+	result = append(result, appBundles.Bundles.Data...)
+
+	for appBundles.GetBundlesEndCursor() != "" {
```

Comment:
I don't know how this will reduce the code duplication. The only difference will be that on line 80 there will be `for shouldFetch := true; shouldFetch; shouldFetch = (appBundles.GetBundlesEndCursor() != "")` but everything else will be the same. 
P.S. The reason I'm setting the value of the `appBundles` to its initial value(the same goes for the `gqlRes`) is to be sure that after `c.runWithTenant(...)` in `appBundles` and `gqlRes` there will be the result of the current graphql call and not the previous one.

Reply:
By using this pattern we can remove the initial call that is before the loop. This pattern works like do while in other languages. So by using it we will have the request in one place only(inside the loop). After the first execution of the request the shouldFetch variable will change to true or false based on the endCursor and the loop will continue if needed